app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleSetUI.kt

```diff
@@ -0,0 +1,161 @@
+          "${base}: ${it}"
+        } ?: base
+        Text(
+          text = text,
+          fontSize = 18.sp
+        )
+      },
+      elevation = 1.dp,
+      navigationIcon = {
+        IconButton(onClick = { onBack() }) {
+          Icon(imageVector = Icons.Default.ArrowBack, contentDescription = null)
+        }
+      },
+      backgroundColor = MaterialTheme.colors.surface
+    )
+
+    state.puzzles?.let { puzzles ->
+      val gridState = rememberLazyGridState()
+      LazyVerticalGrid(
+        state = gridState,
+        columns = GridCells.Adaptive(minSize = 136.dp),
+        modifier = Modifier.fillMaxHeight()
+      ) {
+        items(items = puzzles) {
+          val solved = !state.solutions[it.id].isNullOrEmpty()
+          Surface(
+            shape = MaterialTheme.shapes.medium,
+            modifier = Modifier
+              .fillMaxWidth()
+              .padding(horizontal = 8.dp, vertical = 4.dp)
+          ) {
+            Column(modifier = Modifier
+              .clickable { onPuzzle(it) }
+              .padding(horizontal = 10.dp, vertical = 10.dp)) {
+              Row {
+                it.puzzle.let {
+                  val pos = RulesManager.buildPos(
+                    moves = emptyList(),
+                    boardWidth = it.width,
+                    boardHeight = it.height,
+                    whiteInitialState = it.initial_state.white.toCoordinateSet(),
+                    blackInitialState = it.initial_state.black.toCoordinateSet()
+                  )
+                  Board(
+                    boardWidth = it.width,
+                    boardHeight = it.height,
+                    position = pos,
+                    boardTheme = state.boardTheme,
+                    drawCoordinates = false,
+                    interactive = false,
+                    drawShadow = false,
+                    fadeInLastMove = false,
+                    fadeOutRemovedStones = false,
+                    modifier = Modifier
+                      .weight(1f)
+                      .fillMaxWidth()
+                      .clip(MaterialTheme.shapes.small)
+                  )
+                }
+              }
+              Spacer(modifier = Modifier.height(8.dp))
+              Text(
+                text = it.name,
+                style = TextStyle.Default.copy(
+                  fontSize = 16.sp,
+                  fontWeight = FontWeight.Bold
+                )
+              )
+              if (solved) {
+                Image(
+                  painter = painterResource(R.drawable.ic_check_circle),
+                  modifier = Modifier
+                    .padding(start = 18.dp)
+                    .fillMaxWidth()
+                    .align(Alignment.CenterHorizontally),
+                  contentDescription = "Solved"
+                )
+              }
+            }
+          }
+          LaunchedEffect(true) {
+            fetchSolutions(it.id)
```

Comment:
Why are we triggering data loading from here? Are we trying to do it when the item appears on the screen? Won't this cause a large number of requests to be triggered if the user scrolls fast?

Reply:
Yes! It would, I suppose, but the aim is to avoid launching all the requests at once, because doing that on the regular, for hundreds(?) of users, sounds like a way to upset anoek. Making it work by visibility at least minimizes that, but actually yes I would like a nice solution to filtering out the "scrolls past" requests, is there one?